Mathlib/AlgebraicGeometry/EllipticCurve/Jacobian.lean

```diff
@@ -86,240 +86,329 @@ local macro "pderiv_simp" : tactic =>
+  hx.comp X_eq_of_equiv
+
+lemma not_equiv_of_Y_ne {P Q : Fin 3 → R} (hy : P y * Q z ^ 3 ≠ Q y * P z ^ 3) : ¬P ≈ Q :=
+  hy.comp Y_eq_of_equiv
+
+lemma equiv_of_X_eq_of_Y_eq {P Q : Fin 3 → F} (hPz : P z ≠ 0) (hQz : Q z ≠ 0)
+    (hx : P x * Q z ^ 2 = Q x * P z ^ 2) (hy : P y * Q z ^ 3 = Q y * P z ^ 3) : P ≈ Q := by
+  use Units.mk0 _ hPz / Units.mk0 _ hQz
+  simp only [smul_fin3, Units.val_div_eq_div_val, Units.val_mk0, div_pow, mul_comm, mul_div, ← hx,
+    ← hy, mul_div_cancel_right₀ _ <| pow_ne_zero _ hQz, mul_div_cancel_right₀ _ hQz, fin3_def]
+
+lemma equiv_some_of_Z_ne_zero {P : Fin 3 → F} (hPz : P z ≠ 0) :
+    P ≈ ![P x / P z ^ 2, P y / P z ^ 3, 1] :=
+  equiv_of_X_eq_of_Y_eq hPz one_ne_zero
+    (by linear_combination (norm := (matrix_simp; ring1)) -P x * div_self (pow_ne_zero 2 hPz))
+    (by linear_combination (norm := (matrix_simp; ring1)) -P y * div_self (pow_ne_zero 3 hPz))
+
+end Jacobian
 
 section Equation
 
-/-! ### Equations and nonsingularity -/
+/-! ### Weierstrass equations -/
 
+variable (V) in
 /-- The polynomial $W(X, Y, Z) := Y^2 + a_1XYZ + a_3YZ^3 - (X^3 + a_2X^2Z^2 + a_4XZ^4 + a_6Z^6)$
 associated to a Weierstrass curve `W` over `R`. This is represented as a term of type
 `MvPolynomial (Fin 3) R`, where `X 0`, `X 1`, and `X 2` represent $X$, $Y$, and $Z$ respectively. -/
 noncomputable def polynomial : MvPolynomial (Fin 3) R :=
-  X 1 ^ 2 + C W.a₁ * X 0 * X 1 * X 2 + C W.a₃ * X 1 * X 2 ^ 3
-    - (X 0 ^ 3 + C W.a₂ * X 0 ^ 2 * X 2 ^ 2 + C W.a₄ * X 0 * X 2 ^ 4 + C W.a₆ * X 2 ^ 6)
+  X 1 ^ 2 + C V.a₁ * X 0 * X 1 * X 2 + C V.a₃ * X 1 * X 2 ^ 3
+    - (X 0 ^ 3 + C V.a₂ * X 0 ^ 2 * X 2 ^ 2 + C V.a₄ * X 0 * X 2 ^ 4 + C V.a₆ * X 2 ^ 6)
 
-lemma eval_polynomial (P : Fin 3 → R) : eval P W.polynomial =
-    P y ^ 2 + W.a₁ * P x * P y * P z + W.a₃ * P y * P z ^ 3
-      - (P x ^ 3 + W.a₂ * P x ^ 2 * P z ^ 2 + W.a₄ * P x * P z ^ 4 + W.a₆ * P z ^ 6) := by
+lemma eval_polynomial (P : Fin 3 → R) : eval P V.polynomial =
+    P y ^ 2 + V.a₁ * P x * P y * P z + V.a₃ * P y * P z ^ 3
+      - (P x ^ 3 + V.a₂ * P x ^ 2 * P z ^ 2 + V.a₄ * P x * P z ^ 4 + V.a₆ * P z ^ 6) := by
   rw [polynomial]
   eval_simp
 
+lemma eval_polynomial_of_Z_ne_zero {P : Fin 3 → F} (hPz : P z ≠ 0) : eval P W.polynomial / P z ^ 6 =
+    (W.toAffine.polynomial.eval <| Polynomial.C <| P y / P z ^ 3).eval (P x / P z ^ 2) := by
+  linear_combination (norm := (rw [eval_polynomial, Affine.eval_polynomial]; ring1))
+    W.a₁ * P x * P y / P z ^ 5 * div_self hPz + W.a₃ * P y / P z ^ 3 * div_self (pow_ne_zero 3 hPz)
+      - W.a₂ * P x ^ 2 / P z ^ 4 * div_self (pow_ne_zero 2 hPz)
+      - W.a₄ * P x / P z ^ 2 * div_self (pow_ne_zero 4 hPz) - W.a₆ * div_self (pow_ne_zero 6 hPz)
+
+variable (V) in
 /-- The proposition that a point representative $(x, y, z)$ lies in `W`.
 In other words, $W(x, y, z) = 0$. -/
 def Equation (P : Fin 3 → R) : Prop :=
-  eval P W.polynomial = 0
+  eval P V.polynomial = 0
```

Comment:
Do you think `IsOnCurve` would be a better name for this definition than `Equation`? Maybe that's not in scope for this PR...

Reply:
Sure, I plan to rehaul both `Equation` and `Nonsingular` in a separate future PR. Their names are not ideal (e.g. I think `Nonsingular` should be `IsNonsingular` according to style conventions), and their definitions can be generalised to be correct over more rings (e.g. PIDs).